hal/ia32/init.c

```diff
@@ -0,0 +1,255 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * Configuration detection and other initialisation routines
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Andrzej Stalke
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "hal/string.h"
+#include "hal/cpu.h"
+#include "init.h"
+#include "ia32.h"
+#include "include/errno.h"
+#include "tlb.h"
+
+extern unsigned int _end;
+extern syspage_t *syspage;
+hal_config_t hal_config;
+struct {
+	addr_t pageIterator;
+} init_common;
+
+
+static int _hal_addMemEntry(addr_t start, u32 length, u32 flags)
+{
+	u64 end, pageCount;
+	if (hal_config.memMap.count < HAL_MEM_ENTRIES) {
+		end = start;
+		end += length;
+		if ((end % SIZE_PAGE) != 0) {
+			end += SIZE_PAGE;
+		}
+		end &= ~(SIZE_PAGE - 1);
+
+		start &= ~(SIZE_PAGE - 1);
+		pageCount = (end - start) / SIZE_PAGE;
+
+		hal_config.memMap.entries[hal_config.memMap.count].start = start;
+		hal_config.memMap.entries[hal_config.memMap.count].pageCount = (u32)pageCount;
+		hal_config.memMap.entries[hal_config.memMap.count].flags = flags;
+		hal_config.memMap.count += 1;
+		return 0;
+	}
+	else {
+		return 1;
+	}
+}
+
+
+static inline int _hal_findFreePage(page_t *page)
+{
+	int ret;
+	while (init_common.pageIterator < 0xffff0000) {
+		ret = pmap_getPage(page, &init_common.pageIterator);
+		if ((ret != EOK) || ((page->flags & PAGE_FREE) != 0)) {
+			break;
+		}
+	}
+	return ret;
+}
+
+
+static inline int _hal_configMapPage(u32 *pdir, addr_t pa, void *va, int attr)
+{
+	page_t page;
+	addr_t *ptable;
+	int ret = _pmap_enter(pdir, hal_config.ptable, pa, va, attr, NULL, 0);
+	if (ret < 0) {
+		ret = _hal_findFreePage(&page);
+		if (ret == EOK) {
+			ret = _hal_addMemEntry(page.addr, SIZE_PAGE, PAGE_OWNER_KERNEL | PAGE_KERNEL_PTABLE);
+			if (ret == 0) {
+				ptable = (addr_t *)(syspage->hs.ptable + VADDR_KERNEL);
+				ptable[((u32)hal_config.ptable >> 12) & 0x000003ffu] = (page.addr & ~(SIZE_PAGE - 1)) | (PGHD_WRITE | PGHD_PRESENT);
+				hal_tlbInvalidateLocalEntry(hal_config.ptable);
+				hal_memset(hal_config.ptable, 0, SIZE_PAGE);
+				ret = _pmap_enter(pdir, hal_config.ptable, pa, va, attr, &page, 0);
+			}
+		}
+	}
+	return ret;
+}
+
+
+static inline void *_hal_configMapObject(u32 *pdir, addr_t start, void **va, size_t size, int attr)
```

Comment:
```suggestion
static void *_hal_configMapObject(u32 *pdir, addr_t start, void **va, size_t size, int attr)
```
If you remove the `inline` keyword then it will greatly reduce resulting code.
```bash
$ size init.o
```
with `inline`:
```
   text    data     bss     dec     hex filename
   5076       0       0    5076    13d4 init.o
```
without `inline`
```
   text    data     bss     dec     hex filename
   2837       0       0    2837     b15 init.o
```